notes/force-move-implementation.md

```diff
@@ -0,0 +1,178 @@
+
+We will replace the naive implementation above with a single commitment to the data:
+
+```
+mapping(address => bytes32) channelStorageHashes;
+```
+
+We have two key questions to answer (along with a couple of other minor decisions):
+
+1. What should be the composition of the `channelStorageHash` (i.e. in precisely what order/combination should we hash quantities together to form the root hash)
+2. What should be the composition of the `stateHash` (the thing that the participants sign)
+
+There are lots of possibilities here. To decide between them we need to examine exactly how the quantities will be used.
+To do this we will look at both what the methods need from the channel storage and what the methods need from the channels passed in.
+
+### What do the methods need from the channel storage?
+
+Here we look at which parts of the channel storage the methods actually need to read and which parts can be 'deduced' from the context.
+
+Parts that can be 'deduced' do not need to be passed in.
+For example, as ForceMove only works if the channel is open, which implies that all fields apart from `turnNumRecord` are null, we don't need to provide the other fields: if the `channelStorageHash` doesn't match one where all fields are null we know we can't proceed.
+
+- ForceMove
+  - Can deduce all fields apart from `turnNumRecord` (they must all be null)
+  - Needs to read `turnNumRecord`
+- Response
+  - Can deduce `turnNumRecord` (as the response turnNum must be exactly one more)
+  - Needs to read complete `state` and `finalizesAt`
+- RespondFromAlternative
+  - Can deduce `turnNumRecord`
+  - Needs to read `finalizesAt`
+- Refute
+  - Can deduce `challengerAddress` (from recoverSignature)
+  - Needs to read `turnNumRecord`, `finalizesAt`
+- ConcludeFromOpen
+  - Can deduce all fields apart from `turnNumRecord` (they must all be null)
+  - Needs to read `finalizesAt`
+- ConcludeFromChallenge
+  - Needs to read `finalizesAt`
+- PushOutcome (or whatever)
+  - Needs to read `outcome`
+
+From this we get the following properties that we want for the `channelStorageHash`
+
+- We want it to be efficient to prove that an outcome was part of a state.
+- We want it to be efficient to calculate without having to pass in the full `storedState`
+
+### Proposal for `ChannelStorageHash`
+
+The `ChannelStorageHash` is the hash of:
+
+- `turnNumRecord`
+- `finalizesAt`
+- `stateHash`
+- `challengerAddress`
+- `outcomeHash`
+
+**Why include the `outcomeHash`?** Although the `outcome` is included in the `state`, we include the `outcomeHash` at the top level of the `channelStorageHash` to make it easier for the PushOutcome method to prove what the outcome of the channel was. The tradeoff here is that the methods need to make sure they have the data to calculate it - which adds at most a `bytes32` to their calldata.
```

Comment:
I think there needs to be some bit of data that corresponds to the outcome in the state. There's currently a bit of ambiguity in precisely what "outcome" means. Here outcome could mean "the piece of data you feed to the view function". (But it could also be `{A: 5, B: 5}` if we're doing traditional pure-style settlement.)

Reply:
We should reserve the term "outcome" for the interface between this layer and the payouts layer, so another term should be found for "the piece of data in the view function"

Reply:
T: We can begin to describe time-dependence via outcome interpreters. There exists many open questions in time dependence in general. You _may_ need to allow time dependent transition functions but not clear as of now.

L: How do we do balance refund / multisig deposits with pure `computeOutcome`? Basically rely on order-presedence of withdrawals and let client observe when a deposit has occurred